tests/src/test/java/alluxio/shell/command/CpCommandTest.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.shell.command;
+
+import alluxio.AlluxioURI;
+import alluxio.client.ReadType;
+import alluxio.client.file.FileInStream;
+import alluxio.client.file.URIStatus;
+import alluxio.client.file.options.OpenFileOptions;
+import alluxio.exception.AlluxioException;
+import alluxio.shell.AbstractAlluxioShellTest;
+import alluxio.shell.AlluxioShellUtilsTest;
+
+import com.google.common.io.Closer;
+import org.apache.commons.io.IOUtils;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.IOException;
+
+/**
+ * Tests for cp command.
+ */
+public final class CpCommandTest extends AbstractAlluxioShellTest {
+
+  /**
+   * Tests recursively copying a directory.
+   */
+  @Test
+  public void copyDirTest() throws Exception {
+    AlluxioShellUtilsTest.resetFileHierarchy(mFileSystem);
+    int ret = mFsShell.run("cp", "/testWildCards", "/copy");
+    Assert.assertEquals(0, ret);
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy")));
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy/bar")));
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy/bar/foobar3")));
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy/foo")));
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy/foo/foobar1")));
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy/foo/foobar2")));
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy/foobar4")));
+
+    Assert.assertTrue(
+        equals(new AlluxioURI("/copy/bar/foobar3"), new AlluxioURI("/testWildCards/bar/foobar3")));
+    Assert.assertTrue(
+        equals(new AlluxioURI("/copy/foo/foobar1"), new AlluxioURI("/testWildCards/foo/foobar1")));
+    Assert.assertTrue(
+        equals(new AlluxioURI("/copy/foo/foobar2"), new AlluxioURI("/testWildCards/foo/foobar2")));
+    Assert.assertTrue(
+        equals(new AlluxioURI("/copy/foobar4"), new AlluxioURI("/testWildCards/foobar4")));
+  }
+
+  /**
+   * Tests copying a list of files specified through a wildcard expression.
+   */
+  @Test
+  public void copyWildcardDirTest() throws Exception {
+    AlluxioShellUtilsTest.resetFileHierarchy(mFileSystem);
+    int ret = mFsShell.run("cp", "/testWildCards/*/foo*", "/copy");
+    Assert.assertEquals(0, ret);
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy")));
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy/foobar1")));
```

Comment:
that is the nature of wildcards ... after the wildcard expansion the behavior will be equivalent to

`cp src1 src2 src3 ... dst` which will not preserve the directory structure either